### PR TITLE
Fix workload would be created and deleted indefinitely

### PR DESCRIPTION
**Pull request by kerthcet**
In https://github.com/kubernetes-sigs/kueue/pull/317, we will default workload resource requests to limits if requests not specified. But this will introduce a bug as in job controller reconciling, job will be judged as not equal to workload, so workload will be deleted and re-create indefinitely

#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kueue/issues/590

### CODE REVIEW

**Comment by kerthcet** [comment]
cc @alculquicondor

File: pkg/util/api/api.go [hunk 1]
```diff
@@ -38,3 +42,32 @@ func truncateMessage(message string, limit int) string {
...
+// so we need a function like this to set the defaults.
+func SetContainersDefaults(containers []v1.Container) []v1.Container {
+	// This should only happen in tests.
+	// We didn't return nil here for spec.Containers should not be nil,
+	// we also didn't return empty slice directly here for some unit tests expect this to be nil.
+	if len(containers) == 0 {
+		return containers
+	}
+	res := make([]v1.Container, len(containers))
+	for i, c := range containers {
```
Comment (alculquicondor) [inline pkg/util/api/api.go:59]: Let's add a comment that this logic is copied from k8s defaulting https://github.com/kubernetes/kubernetes/blob/ab002db78835a94bd19ce4aaa46fb39b4d9c276f/pkg/apis/core/v1/defaults.go#L144

File: pkg/util/api/api_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (alculquicondor) [inline pkg/util/api/api_test.go:2]: 2023

File: pkg/controller/workload/job/job_controller.go [hunk 3]
```diff
@@ -621,12 +623,12 @@ func jobAndWorkloadEqual(job *batchv1.Job, wl *kueue.Workload) bool {
 
 	// nodeSelector may change, hence we are not checking for
 	// equality of the whole job.Spec.Template.Spec.
-	if !equality.Semantic.DeepEqual(job.Spec.Template.Spec.InitContainers,
-		wl.Spec.PodSets[0].Spec.InitContainers) {
+	initContainersShadowCopy := utilapi.SetContainersDefaults(job.Spec.Template.Spec.InitContainers)
```
Comment (alculquicondor) [inline pkg/controller/workload/job/job_controller.go:626]: nit
```suggestion
	initContainersDefalted := utilapi.SetContainersDefaults(job.Spec.Template.Spec.InitContainers)
```

File: pkg/util/api/api.go [hunk 4]
```diff
@@ -16,6 +16,10 @@ limitations under the License.
 
 package api
```
Comment (alculquicondor) [inline pkg/util/api/api.go:17]: I'm not convinced about putting this function here. But we can revisit once we have the jobs library completed, which will probably use it.

File: pkg/util/api/api.go [hunk 5]
```diff
@@ -38,3 +42,32 @@ func truncateMessage(message string, limit int) string {
...
 }
+
+// SetContainersDefaults will fill up the container defaults.
+// Note that this is used to bridge the gap between job and workload,
+// E.g. job may have resources limits configured but no requests,
+// this is not allowed in workload but we can't update the job directly,
+// so we need a function like this to set the defaults.
+func SetContainersDefaults(containers []v1.Container) []v1.Container {
+	// This should only happen in tests.
+	// We didn't return nil here for spec.Containers should not be nil,
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/util/api/api.go:53]: As mentioned, this only happens in test, If we return nil here, integration tests will fail because it shouldn't be nil.

If we return empty slice here, unit tests will fail like https://github.com/kubernetes-sigs/kueue/blob/271e66a6441f6fd96f63d6e62f9d39b723da30bb/apis/kueue/webhooks/workload_webhook_test.go#L43-L50

I think I should polish the comments.

File: pkg/util/api/api.go [hunk 6]
```diff
@@ -38,3 +42,32 @@ func truncateMessage(message string, limit int) string {
 	suffix := " ..."
 	return message[:limit-len(suffix)] + suffix
 }
+
+// SetContainersDefaults will fill up the container defaults.
+// Note that this is used to bridge the gap between job and workload,
+// E.g. job may have resources limits configured but no requests,
+// this is not allowed in workload but we can't update the job directly,
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/util/api/api.go]: Updated. One more thing, if LimitRange is introduced, workload with no requests should be rejected, wdyt?

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/util/api/api.go]: Didn't we vote against that already? #485

**Comment by alculquicondor** [comment]
Instead, can we focus on the library?

**Comment by mimowo** [comment]
> Instead, can we focus on the library?

Yes, I was just thinking it makes sense to get rid of the issue as it is now fresh in our minds, and the fix should be just a couple of lines, but I agree with the priorities.

**Comment by kerthcet** [comment]
Let me rebase this one until https://github.com/kubernetes-sigs/kueue/pull/616 got merged.

**Comment by kerthcet** [comment]
Rebased.

**Comment by alculquicondor** [comment]
@kerthcet wdyt about https://github.com/kubernetes-sigs/kueue/pull/600#issuecomment-1462272626?
I wonder if we should hold for this PR given that it would change significantly.

**Comment by kerthcet** [comment]
/close

superseded by https://github.com/kubernetes-sigs/kueue/pull/600
